### ISSUE TITLE
Assistant API - toolCall event not fired for file_search

### ISSUE THREAD

**Issue by magic-maker** [openai/openai-node#850]
### Confirm this is a Node library issue and not an underlying OpenAI API issue

- [X] This is an issue with the Node library

### Describe the bug

The helper events "toolCall" (created, delta, done) with a message attachments with file and tool type "file_search" are never fired.
It works well for "code_interpreter" type.

### To Reproduce

Just attach a toolCall event to a run with a message attachment type "file_search" and a file_id from a vector store.

### OS

windows

### Node version

Node v21.5.0

### Library version

openai v4.47.1

**Comment by rattrayalex** [comment]
Thanks for the report, we'll look into this. If you could share a script that reproduces the problem that may be helpful.

**Comment by Eldo420** [comment]
### Brief Overview of the Problem:
The `toolCall` event should be triggered when a tool operation, such as `file_search`, is invoked. However, users are experiencing that this specific event is not firing as expected.

### Steps to Troubleshoot and Resolve:

1. **Reproduce the Issue**
   - Create a sample script to reproduce the problem. This will help ensure that the issue is consistent and not due to external factors.

2. **Check Event Listener Setup**
   - Verify that the event listener for `toolCall` is correctly set up in your application.

3. **Inspect the Event Dispatching Logic**
   - Dive into the source code to identify where events are dispatched and ensure `file_search` is included.

4. **Update the Codebase if Necessary**
   - If the `file_search` tool is not triggering the event, update the codebase to include this functionality.
   
5. **Add Tests to Validate the Fix**
   - Create tests to ensure the `toolCall` event fires for `file_search` and other tools.

### Proposed Solution

#### Example Script to Reproduce the Issue

```javascript
const { OpenAIApi } = require('openai');

const openai = new OpenAIApi({
  apiKey: 'your-api-key-here'
});

const onToolCall = (event) => {
  console.log("Tool Call Event Fired:", event);
  // Additional assertions/validations can be added here to confirm event details.
};

(async () => {
  try {
    // Example function to simulate file_search operation
    openai.on('toolCall', onToolCall);

    // Simulate a file_search tool call
    const response = await openai.file_search({
      // Provide necessary parameters for file_search
      query: 'example-query'
    });

    console.log('File Search Response:', response);
  } catch (error) {
    console.error('Error during file_search:', error);
  }
})();
```

#### Checking Event Listener Setup

Ensure that the `on('toolCall', callback)` event listener is correctly set for capturing the `toolCall` event.

#### Inspect and Update Event Dispatching Logic

Go to the part of the codebase where events are dispatched:

```javascript
// pseudo code: event_dispatcher.js or similar file
dispatch(eventType, eventData) {
  if (this.listeners[eventType]) {
    this.listeners[eventType].forEach(callback => callback(eventData));
  }
}
```

Ensure `file_search` tool calls this dispatch method with appropriate parameters.

#### Example Update in Codebase

```javascript
// pseudo code: file_search.js or similar file
fileSearch(params) {
  // Perform the file search operation...

  // After the operation, dispatch the toolCall event
  this.dispatch('toolCall', { tool: 'file_search', params, result });

  return result;
}
```

#### Adding Tests

Add tests to validate that the `toolCall` event is indeed fired for `file_search`.

```javascript
describe('Tool Call Event Tests', () => {
  it('should fire toolCall event for file_search', (done) => {
    const openai = new OpenAIApi({ apiKey: 'your-api-key-here' });

    openai.on('toolCall', (event) => {
      try {
        expect(event.tool).toEqual('file_search');
        done();
      } catch (error) {
        done(error);
      }
    });

    openai.file_search({ query: 'example-query' });
  });
});
```

### Implementation Plan:

1. **Reproduce the Issue**:
   - Use the provided script to consistently reproduce the issue.

2. **Check Event Listener**:
   - Validate that the event listener is set correctly and debug any issues around it.

3. **Inspect Event Dispatching**:
   - Inspect the codebase to ensure the `file_search` tool triggers the `toolCall` event.

4. **Apply Fixes**:
   - Modify the codebase to ensure `file_search` also triggers the `toolCall` event when invoked.

5. **Add and Run Tests**:
   - Add tests to verify the `toolCall` event fires for `file_search` and all other tools, then run the tests.

6. **Update Repository and Documentation**:
   - Commit changes and update documentation to reflect the event firing for `file_search`.

7. **Close the Issue**:
   - Once the changes are validated and merged, close issue #850 with a note explaining the fix.

### Example Commit Message:

```
fix: Ensure toolCall event fires for file_search operation

- Updated event dispatch logic to include file_search operations.
- Added tests to verify toolCall event firing for file_search.
- Closes #850.
```

By following these steps, we can resolve the `toolCall` event not firing for the `file_search` tool, ensuring consistent behavior across all tools within the Assistant API.

**Comment by rattrayalex** [comment]
Closing as a duplicate of https://github.com/openai/openai-node/issues/771
